backend/src/main/java/eu/clarin/switchboard/resources/ToolsResource.java

```diff
@@ -42,4 +46,46 @@ public Response getTools(@QueryParam("mediatype") String mediatype,
         tools.sort((t1, t2) -> t1.getName().compareToIgnoreCase(t2.getName()));
         return Response.ok(tools).build();
     }
+
+    @GET
+    @Path("logos/{logoName}")
+    public Response getTools(@PathParam("logoName") String logoName) throws IOException {
+        String imageMimeType = URLConnection.guessContentTypeFromName(logoName);
+        java.nio.file.Path logo = Paths.get(toolConfig.getLogoRegistryPath(), logoName);
+
+        byte[] data;
+        if (logo.toFile().exists()) {
+            data = Files.readAllBytes(logo);
+        } else {
+            LOGGER.debug("logo not found, trying to find it in resources: " + logoName);
+            data = this.readSmallResource(logoName);
+        }
+
+        if (data == null) {
+            return Response.status(Response.Status.NOT_FOUND).build();
+        }
+
+        StreamingOutput fileStream = output -> {
+            output.write(data);
+            output.flush();
+        };
+        return Response
+                .ok(fileStream, imageMimeType)
+                .build();
+    }
+
+    private byte[] readSmallResource(String name) throws IOException {
+        InputStream is = this.getClass().getResourceAsStream("/webui/images/" + name);
+        if (is == null) {
+            return null;
+        }
+        byte[] buf = new byte[1024 * 1024];
+        int count = is.read(buf);
+        is.close();
+        if (count > 0) {
+            return Arrays.copyOfRange(buf, 0, count);
+        } else {
+            return new byte[0];
+        }
```

Comment:
_might_ be tidier to use a ByteArrayOutputStream and something like Guava's [ByteStreams.copy](https://guava.dev/releases/19.0/api/docs/com/google/common/io/ByteStreams.html#copy(java.io.InputStream,%20java.io.OutputStream))? 

Also the InputStream can be used more safely using a try statement ([try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)]

Reply:
I agree, it was a sloppy implementation